components/ai_chat/browser/ai_chat_database.cc

```diff
@@ -0,0 +1,271 @@
+    return false;
+  }
+
+  sql::Statement insert_conversation_entry_statement(GetDB().GetUniqueStatement(
+      "INSERT INTO conversation_entry(" CONVERSATION_ENTRY_FIELDS
+      ") VALUES(NULL, ?, ?, ?)"));
+  insert_conversation_entry_statement.BindTimeDelta(
+      0, SerializeTimeToDelta(entry->date));
+  insert_conversation_entry_statement.BindInt(
+      1, static_cast<int>(entry->character_type));
+  insert_conversation_entry_statement.BindInt64(
+      2, get_conversation_id_statement.ColumnInt64(0));
+
+  if (!insert_conversation_entry_statement.Run()) {
+    DVLOG(0) << "Failed to execute 'conversation_entry' insert statement\n";
+    return false;
+  }
+
+  int64_t conversation_entry_row_id = GetDB().GetLastInsertRowId();
+
+  for (const auto& text : entry->texts) {
+    sql::Statement insert_conversation_text_statement(
+        GetDB().GetUniqueStatement(
+            "INSERT INTO "
+            "conversation_entry_text(" CONVERSATION_ENTRY_TEXT_FIELDS
+            ") VALUES(NULL, ?, ?)"));
+
+    insert_conversation_text_statement.BindString(0, text->text);
+    insert_conversation_text_statement.BindInt64(1, conversation_entry_row_id);
+
+    if (!insert_conversation_text_statement.Run()) {
+      DVLOG(0)
+          << "Failed to execute 'conversation_entry_text' insert statement\n";
+      return false;
+    }
+  }
+
+  transaction.Commit();
+  return true;
+}
+
+bool AIChatDatabase::DeleteConversation(int64_t conversation_id) {
+  sql::Transaction transaction(&db_);
+  if (!transaction.Begin()) {
+    DVLOG(0) << "Transaction cannot begin\n";
+    return false;
+  }
+
+  sql::Statement delete_conversation_statement(
+      GetDB().GetUniqueStatement("DELETE FROM conversation"
+                                 "WHERE id=?"));
+  delete_conversation_statement.BindInt64(0, conversation_id);
+  delete_conversation_statement.Run();
+
+  sql::Statement select_conversation_entry_statement(
+      GetDB().GetUniqueStatement("SELECT id FROM conversation_entry"
+                                 "WHERE conversation_id=?"));
+
+  sql::Statement delete_conversation_text_statement(
+      GetDB().GetUniqueStatement("DELETE FROM conversation_entry_text"
+                                 "WHERE conversation_entry_id=?"));
+
+  // We remove all conversation text associated to |conversation_id|
+  while (select_conversation_entry_statement.Step()) {
+    delete_conversation_text_statement.BindInt64(
+        0, select_conversation_entry_statement.ColumnInt64(0));
+    delete_conversation_text_statement.Run();
+  }
+
+  sql::Statement delete_conversation_entry_statement(
+      GetDB().GetUniqueStatement("DELETE FROM conversation_entry"
+                                 "WHERE conversation_entry_id=?"));
+  delete_conversation_entry_statement.BindInt64(0, conversation_id);
+  // At last we remove all conversation entries associated to |conversation_id|
+  delete_conversation_entry_statement.Run();
+
+  transaction.Commit();
```

Comment:
I think we just need to let it go out of scope and not trigger .Commit explicitly, so that the implicit rollback happens when it goes out of scope.

Changed.

```
// Explicitly rolls back the transaction. All changes will be forgotten.
  //
  // Most features can avoid calling this method, because Transactions that do
  // not get Commit()ed are automatically rolled back when they go out of scope.
  //
  // This method is not idempotent. Calling Rollback() twice on a Transaction
  // will cause a DCHECK crash.
  //
  // Must be called after a successful call to Begin(). Must not be called after
  // Commit().
  void Rollback();
```